src/signalrclient/hub_connection_impl.cpp

```diff
@@ -151,25 +154,125 @@ namespace signalr
                         return;
                     }
 
-                    if (exception)
+                    try
                     {
-                        callback(exception);
-                        return;
+                        if (exception == nullptr)
+                        {
+                            connection->m_handshakeTask->get();
+                            {
+                                std::lock_guard<std::mutex> lock(*handshake_request_lock);
+                                assert(*handshake_callback_done == false);
+                                *handshake_callback_done = true;
+                            }
+                            callback(nullptr);
+                        }
+                    }
+                    catch (...)
+                    {
+                        exception = std::current_exception();
                     }
 
-                    try
+                    if (exception != nullptr)
                     {
-                        connection->m_handshakeTask->get();
-                        callback(nullptr);
+                        {
+                            std::lock_guard<std::mutex> lock(*handshake_request_lock);
+                            *handshake_callback_done = true;
+                        }
+
+                        connection->m_connection->stop([callback, exception](std::exception_ptr)
+                            {
+                                callback(exception);
+                            }, nullptr);
                     }
-                    catch (...)
+                };
+
+                std::shared_ptr<bool> handshake_request_done = std::make_shared<bool>();
+                auto handshake_request = handshake::write_handshake(connection->m_protocol);
+                auto handshake_task = connection->m_handshakeTask;
+                auto handshake_timeout = connection->m_signalr_client_config.get_handshake_timeout();
+
+                connection->m_disconnect_cts->register_callback([handle_handshake, handshake_request_lock, handshake_callback_done, handshake_request_done]()
+                    {
+                        bool run_callback = false;
+                        {
+                            std::lock_guard<std::mutex> lock(*handshake_request_lock);
+                            // no op after connection started successfully
+                            if (*handshake_callback_done == true)
+                            {
+                                return;
+                            }
+
+                            // if the request isn't completed then no one is waiting on the handshake task
+                            // so we need to run the callback here instead of relying on connection.send completing
+                            if (*handshake_request_done == false)
+                            {
+                                *handshake_callback_done = true;
```

Comment:
Can you get by with just handshake_callback_done or is there some reason handshake_request_done needs to be tracked independently?

Reply:
I tried with just one initially but switched to two values for simplicity. There are 3 things going on here, connection.send taking too long, handshake response taking too long, and connection.stop being called.

Maybe I can fix this by exposing `handshake_task.is_set()` or something, will give it a look.